$ \newcommand{\bra}[1]{\langle #1|} $
$ \newcommand{\ket}[1]{|#1\rangle} $
$ \newcommand{\braket}[2]{\langle #1|#2\rangle} $
$ \newcommand{\dot}[2]{ #1 \cdot #2} $
$ \newcommand{\biginner}[2]{\left\langle #1,#2\right\rangle} $
$ \newcommand{\mymatrix}[2]{\left( \begin{array}{#1} #2\end{array} \right)} $
$ \newcommand{\myvector}[1]{\mymatrix{c}{#1}} $
$ \newcommand{\myrvector}[1]{\mymatrix{r}{#1}} $
$ \newcommand{\mypar}[1]{\left( #1 \right)} $
$ \newcommand{\mybigpar}[1]{ \Big( #1 \Big)} $
$ \newcommand{\sqrttwo}{\frac{1}{\sqrt{2}}} $
$ \newcommand{\dsqrttwo}{\dfrac{1}{\sqrt{2}}} $
$ \newcommand{\onehalf}{\frac{1}{2}} $
$ \newcommand{\donehalf}{\dfrac{1}{2}} $
$ \newcommand{\hadamard}{ \mymatrix{rr}{ \sqrttwo & \sqrttwo \\ \sqrttwo & -\sqrttwo }} $
$ \newcommand{\vzero}{\myvector{1\\0}} $
$ \newcommand{\vone}{\myvector{0\\1}} $
$ \newcommand{\stateplus}{\myvector{ \sqrttwo \\  \sqrttwo } } $
$ \newcommand{\stateminus}{ \myrvector{ \sqrttwo \\ -\sqrttwo } } $
$ \newcommand{\myarray}[2]{ \begin{array}{#1}#2\end{array}} $
$ \newcommand{\X}{ \mymatrix{cc}{0 & 1 \\ 1 & 0}  } $
$ \newcommand{\I}{ \mymatrix{rr}{1 & 0 \\ 0 & 1}  } $
$ \newcommand{\Z}{ \mymatrix{rr}{1 & 0 \\ 0 & -1}  } $
$ \newcommand{\Htwo}{ \mymatrix{rrrr}{ \frac{1}{2} & \frac{1}{2} & \frac{1}{2} & \frac{1}{2} \\ \frac{1}{2} & -\frac{1}{2} & \frac{1}{2} & -\frac{1}{2} \\ \frac{1}{2} & \frac{1}{2} & -\frac{1}{2} & -\frac{1}{2} \\ \frac{1}{2} & -\frac{1}{2} & -\frac{1}{2} & \frac{1}{2} } } $
$ \newcommand{\CNOT}{ \mymatrix{cccc}{1 & 0 & 0 & 0 \\ 0 & 1 & 0 & 0 \\ 0 & 0 & 0 & 1 \\ 0 & 0 & 1 & 0} } $
$ \newcommand{\norm}[1]{ \left\lVert #1 \right\rVert } $
$ \newcommand{\pstate}[1]{ \lceil \mspace{-1mu} #1 \mspace{-1.5mu} \rfloor } $
$ \newcommand{\greenbit}[1] {\mathbf{{\color{green}#1}}} $
$ \newcommand{\bluebit}[1] {\mathbf{{\color{blue}#1}}} $
$ \newcommand{\redbit}[1] {\mathbf{{\color{red}#1}}} $
$ \newcommand{\brownbit}[1] {\mathbf{{\color{brown}#1}}} $
$ \newcommand{\blackbit}[1] {\mathbf{{\color{black}#1}}} $

<font style="font-size:28px;" align="left"><b> Quantum State </b></font>
<br>


_The overall probability must be 1 when we observe a quantum system._

For example, the following vectors <u>cannot</u> be a valid quantum state:

$$
    \myvector{ \dfrac{1}{2} \\ \dfrac{1}{2} }
    \mbox{ and }
    \myvector{ \dfrac{\sqrt{3}}{2} \\ \dfrac{1}{\sqrt{2}} }.
$$

For the first vector, the probabilities of observing the states $\ket{0} $ and $ \ket{1} $ are $ \dfrac{1}{4} $.

So, the overall probability of getting a result is $ \dfrac{1}{4} + \dfrac{1}{4} = \dfrac{1}{2} $, which is less than 1.

For the second vector, the probabilities of observing the states $\ket{0} $ and $ \ket{1} $ are respectively $ \dfrac{3}{4} $ and $ \dfrac{1}{2} $.

So, the overall probability of getting a result is $ \dfrac{3}{4} + \dfrac{1}{2} = \dfrac{5}{4} $, which is greater than 1.

<font color="blue"><b>The summation of amplitude squares must be 1 for a valid quantum state.</b></font>

<font color="blue"><b>More formally, a quantum state can be represented by a vector having length 1, and vice versa.</b></font>

<i>The summation of amplitude squares gives the square of the length of vector.

But, this summation is 1, and its square root is also 1. So, we can use the term <u>length</u> in the definition.</i>

<i> <b>Technical notes:</b> We represent a quantum state as $ \ket{u} $ instead of $ u $. Remember the relation between the length and dot product: $ \norm{u} = \sqrt{\dot{u}{u}} $.
    
In quantum computation, we use inner product instead of dot product, which is defined on complex numbers. By using bra-ket notation, $ \norm{ \ket{u} } = \sqrt{ \braket{u}{u} } = 1 $, or equivalently $ \braket{u}{u} = 1 $, where $ \braket{u}{u} $ is a short form of $ \bra{u}\ket{u} $. For real-valued vectors, $ \braket{v}{v} = \dot{v}{v}  $.
</i>

###In this assignment, you will learn how to check if a given vector is a valid quantum state.

####**Task 1. Checking if a Vector is Normalized**


1. Write a Python function is_normalized(v) that takes a NumPy array representing a vector v as input and returns a Boolean value indicating if the vector is normalized. A vector is normalized if the sum of the squared absolute values of its elements is equal to 1.

2. Test your function with the following vectors:

a. [1/2, 1/2] <br>
b. [1, 0, 0, 0] <br>
c. [1, 1] <br>
d. [-1/2, 1j/2] <br>
e. [1, 2, 3] <br>

3. Discuss your results. Which vectors are normalized and which are not? Why?

In [6]:
from functools import reduce
import math

def is_normalized(v):
  return reduce(lambda acc, x: acc + (x ** 2), v, 0) == 1

print(f"A = {is_normalized([1/2, 1/2])}")
print(f"B = {is_normalized([1, 0, 0, 0])}")
print(f"C = {is_normalized([1, 1])}")
print(f"D = {is_normalized([-1/2, 1j/2])}")
print(f"E = {is_normalized([1, 2, 3])}")

A = False
B = True
C = False
D = False
E = False


####**Task 2: Checking if a Vector is a Column Vector**
1. Write a Python function is_column_vector(v) that takes a NumPy array representing a vector v as input and returns a Boolean value indicating if the vector is a column vector. A vector is a column vector if it has a shape of (n, 1) or (n,), where n is the number of elements in the vector.

2. Test your function with the following vectors:

a. [1/2, 1/2] <br>
b. [[1], [0], [0], [0]] <br>
c. [1, 1] <br>
d. [-1/2, 1j/2] <br>
e. [1, 2, 3] <br>

3. Discuss your results. Which vectors are column vectors and which are not? Why?

In [10]:
import numpy as np

def is_column_vector(v):
    if isinstance(v, np.ndarray):
        return v.ndim == 1 or (v.ndim == 2 and v.shape[1] == 1)
    return False


print(f"A = {is_column_vector(np.array([1/2, 1/2]))}")
print(f"B = {is_column_vector(np.array([[1], [0], [0], [0]]))}")
print(f"C = {is_column_vector(np.array([1,1]))}")
print(f"D = {is_column_vector(np.array([-1/2, 1/2j]))}")
print(f"E = {is_column_vector(np.array([1, 2, 3]))}")

(2,)
A = True
(4, 1)
B = True
(2,)
C = True
(2,)
D = True
(3,)
E = True


###Import qiskit and other necessary libraries

Check if state $$ vec = \dfrac{1}{\sqrt{2}}\myvector{1 \\ 0  \\ i \\ 0 }$$
is a valid quantum state.<br><br>
Initialize a quantum state by vector $vec$ using $Statevector$ function in qiskit.

In [24]:
# !pip3 install qiskit
# !pip3 install qiskit-aer
import qiskit
from qiskit import QuantumRegister, ClassicalRegister, QuantumCircuit
from qiskit.quantum_info import Statevector
from qiskit_aer import Aer
import numpy as np

vec = Statevector([1, 0 , 1j, 0])/math.sqrt(2)
print(np.vdot(vec, vec))

(0.9999999999999998+0j)


code shows a slight numerical error due to precision, but if calculated manually, the vector does represent a valid quantum state |(1/√2)|^2 + |(i/√2)|^2 = 1/2 + 1/2 = 1



<h3> Task 3 </h3>

If the following vectors are valid quantum states defined with real numbers, then what can be the values of $a$ and $b$?

$$
    \ket{v} = \myrvector{a \\ -0.1 \\ -0.3 \\ 0.4 \\ 0.5}
    ~~~~~ \mbox{and} ~~~~~
   \ket{u} = \myrvector{ \frac{1}{\sqrt{2}} \\ \frac{1}{\sqrt{b}} \\ -\frac{1}{\sqrt{3}} }.
$$

1. 
|a|^2 + |-0.1|^2 + |-0.3|^2 + |0.4|^2 + |0.5|^2 = 1 <br>
a^2 = 1 - 1/100 - 9/100 - 16/100 - 25/100 <br>
a^2 = 1 - 51/100 <br>
a^2 = 49/100 <br>
a = 7/10 v a = -7/10 <br>

2.
|1/√2|^2 + |1/√b|^2 + |-1/√3|^2 = 1 <br>
1/b = 1 - 1/2 - 1/3 <br>
1/b = 1/6 <br>
b = 6 <br>



<h3> Quantum Operators </h3>

Once the quantum state is defined, the definition of quantum operator is very easy.

<font color="blue"><b>Any length preserving (square) matrix is a quantum operator, and vice versa.</b></font>

<a id="task4"></a>
<h3> Task 4</h3>

The following matrix is called Hadamard operator:

$$
    H = \hadamard.
$$

Randomly create a 2-dimensional quantum state, and test whether Hadamard operator preserves its length or not.

<b>Write a function</b> that returns a randomly created 2-dimensional quantum state.

<i>Hint:
<ul>
    <li> Pick two random values between -100 and 100 for the amplitudes of state 0 and state 1 </li>
    <li> Find an appropriate normalization factor to divide each amplitude such that the length of quantum state should be 1 </li>
</ul>
</i>

<b>Write a function</b> that determines whether a given vector is a valid quantum state or not.

(Due to precision problem, the summation of squares may not be exactly 1 but very close to 1, e.g., 0.9999999999999998.)

<b>Repeat 10 times:</b>
<ul>
    <li> Randomly pick a quantum state </li>
    <li> Check whether the picked quantum state is valid </li>
    <li> Multiply Hadamard operator with the randomly created quantum state </li>
    <li> Check whether the quantum state in result is valid </li>
</ul>

In [6]:
#
# you may define your first function in a separate cell
#

from random import randrange
import numpy as np

def random_quantum_state():
    alpha = np.random.uniform(-100, 100)
    beta = np.random.uniform(-100, 100)

    norm_factor = np.sqrt(alpha**2 + beta**2)
    alpha_normalized = alpha / norm_factor
    beta_normalized = beta / norm_factor

    return np.array([alpha_normalized, beta_normalized])

def is_valid_quantum_state(state):
    norm = np.abs(state[0])**2 + np.abs(state[1])**2
    return np.isclose(norm, 1, atol=1e-10)

def apply_hadamard(state):
    H = 1/np.sqrt(2) * np.array([[1, 1], [1, -1]])
    return np.dot(H, state)


for i in range (0, 10):
    state = random_quantum_state()
    print(f"State {state}")
    print(f"Is valid: {is_valid_quantum_state(state)}")
    print(f"Is valid after applying H: {is_valid_quantum_state(apply_hadamard(state))}")

State [ 0.85929544 -0.51147957]
Is valid: True
Is valid after applying H: True
State [-0.74832935 -0.66332735]
Is valid: True
Is valid after applying H: True
State [-0.88673771  0.4622729 ]
Is valid: True
Is valid after applying H: True
State [0.73988866 0.67272934]
Is valid: True
Is valid after applying H: True
State [-0.71870314  0.69531705]
Is valid: True
Is valid after applying H: True
State [-0.60972076  0.7926163 ]
Is valid: True
Is valid after applying H: True
State [-0.71693801  0.69713693]
Is valid: True
Is valid after applying H: True
State [-0.68926329  0.72451095]
Is valid: True
Is valid after applying H: True
State [-0.42812639 -0.90371887]
Is valid: True
Is valid after applying H: True
State [-0.59166789  0.80618181]
Is valid: True
Is valid after applying H: True


In [ ]:
#
# your code is here
#


<a id="task5"></a>
<h3> Task 5</h3>

Let $ \ket{u} =  \myvector{x \\ y} \in \mathbb{R}^2 $ is a quantum state.

Show that $ \ket{u'} = H \ket{u} $ is also a quantum state.

_Hint: The length of $ \ket{u'} $ must be 1, given that the length of $ \ket{u} $ is 1._

|u> = [a, b]
sqrt(a ** 2 + b ** 2) = 1

p = 1/sqrt(2)

H|u> = [
  p * a + p * b,
  p * a - p * b
]

sqrt( (pa + pb) ** 2 + (p*a - p*b) ** 2) =  <br>
sqrt(p**2 * (a + b) ** 2 + p ** 2 (a - b) ** 2) =  <br>
sqrt(p**2 * (2a ** 2 + 2b ** 2)) = <br>
sqrt(2p**2 * (a ** 2 + b ** 2)) = <br>
sqrt(2p ** 2 * 1) = <br>
sqrt(2p ** 2) = <br>
sqrt (2 * 1/2) = <br>
sqrt (1) =  <br>
1